README.md
# Unfinished Haskell bindings to LIFX LAN protocol (v2.0)

Status: Haitus. This was primarily a POC for someone else who, ultimately,
found a completed node library doing the same thing. I am publicising this
solely in case someone else decides to take this from me (I do own some LIFX,
so I might also return to this). The whole thing was under 3 hours of work, I
would treat this repository as one unfinished perspective on how to set up
this kind of project.

Features:
- All of documentation (02/19/2019) has been mapped into datatypes.
- Uses serialise/cborg under the hood because _cborg has a readable binary
  format_ and you can convert the cbor spec to JSON easily (this would make an
  HTTP api trivial, in theory).
- Intended to use udp-streaming since the `streaming` libraries can be easily
  converted into pipes (and maybe conduit) from what I've been told.

TODO:
- [ ] Datatypes other than the header types need `serialise` instances.
- [ ] Need to integrate with udp-streaming
- [ ] Need to expose an HTTP API since there doesn't seem to be too many
      haskellers working with lifx.



ChangeLog.md
# Changelog for lifx

## Unreleased changes


src/Data/LIFX/Tile.hs
-------------------------------------------------------------------------------
-- |
-- Module    :  Data.LIFX.Tile
-- Copyright :  (c) Sam Stites 2017
-- License   :  BSD-3-Clause
-- Maintainer:  fnz@fgvgrf.vb (cipher:rot13)
-- Stability :  experimental
-- Portability: non-portable
--
-- Tile Messages
--
-- These messages are specific to controlling and querying LIFX Tile devices.
-- See the page on Tiles for more information about these products.
-------------------------------------------------------------------------------
module Data.LIFX.Tile where

import Prelude
import GHC.Word
import GHC.Int
import Data.ByteString

import Data.LIFX.Light

-- * Message Field Data Types

-- The device_version fields are the same as those in a Device::StateVersion
-- message and the the firmware_build and firmware_version fields are the same
-- as those in a Device::StateHostFirmware message.
--
-- The width and height fields are a number of pixels that are on each axis of
-- the tile.
--
-- The user_x and user_y fields are used to record the position of each tile.
-- See the page on Tile Control for more information about user_x and user_y
--
-- The accel_meas fields are the gravity measurements on X,Y and Z axes. The
-- measurement range is ±2g with a sensitivity of 4096 LSB/g.
--
-- The following javascript is an example of determining the rotation of the
-- device from those measurements.
--
-- > function determine_rotation(x, y, z) {
-- >   var absX = abs(x)
-- >   var absY = abs(y)
-- >   var absZ = abs(z)
-- >
-- >   if (x == -1 && y == -1 && z == -1) {
-- >     // Invalid data, assume right-side up.
-- >     return "rightSideUp"
-- >
-- >   } else if (absX > absY && absX > absZ) {
-- >     if (x > 0) {
-- >       return "rotateRight"
-- >     } else {
-- >       return "rotateLeft"
-- >     }
-- >
-- >   } else if (absZ > absX && absZ > absY) {
-- >     if (z > 0) {
-- >       return "faceDown"
-- >     } else {
-- >       return "faceUp"
-- >     }
-- >
-- >   } else {
-- >     if (y > 0) {
-- >       return "upsideDown"
-- >     } else {
-- >       return "rightSideUp"
-- >     }
-- >   }
-- > }
data Tile = Tile
  { accel_meas_x :: Int16
  , accel_meas_y :: Int16
  , accel_meas_z :: Int16
  , reserved     :: Int16
  , user_x       :: Float
  , user_y       :: Float
  , width        :: Word8
  , height       :: Word8
  , reserved0    :: Word8
  , device_version_vendor  :: Word32
  , device_version_product :: Word32
  , device_version_version :: Word32
  , firmware_build         :: Word64
  , reserved1              :: Word64
  , firmware_version       :: Word32
  , reserved2              :: Word32
  } deriving (Eq, Ord, Show)

-- * Tile Messages

data TileMessages
  = GetDeviceChain -- ^ 701
  -- This message returns information about the tiles in the chain. It responds with a StateDeviceChain message.

  | StateDeviceChain Word8 [Tile] Word8 -- ^ 702
  -- Field             | Type
  -- start_index       | unsigned 8-bit integer
  -- tile_devices[16]  | 16 Tile
  -- total_count       | unsigned 8-bit integer

  | SetUserPosition Word8 Word16 Float Float -- ^ 703
  -- Used to tell each tile what their position is.
  --
  -- Field      | Type
  -- tile_index | unsigned 8-bit integer
  -- reserved   | unsigned 16-bit integer
  -- user_x     | 32-bit float
  -- user_y     | 32-bit float
  --
  -- The tile_index is a 0 based index used to address a particular tile in the chain.
  --
  -- The user_x and user_y fields are used to record the position of each tile.
  -- See the page on Tile Control for more information about user_x and user_y.
  --
  -- Warning!
  --
  -- Be very careful when setting these values, as it has the chance to
  -- greatly upset users if you get it wrong. Make sure you have read, and
  -- fully understand the information on the Tile Control page before setting
  -- these values.

  | GetTileState64 Word8 Word8 Word8 Word8 Word8 Word8 -- ^ 707
  -- Get the state of 64 pixels in the tile in a rectangle that has a starting
  -- point and width.
  --
  -- Field       | Type
  -- tile_index  | unsigned 8-bit integer
  -- length      | unsigned 8-bit integer
  -- reserved    | unsigned 8-bit integer
  -- x           | unsigned 8-bit integer
  -- y           | unsigned 8-bit integer
  -- width       | unsigned 8-bit integer
  --
  -- The tile_index is used to control the starting tile in the chain and
  -- length is used to get the state of that many tiles beginning from the
  -- tile_index. This will result in a separate response from each tile.
  --
  -- For the LIFX Tile it really only makes sense to set x and y to zero, and width to 8.

  | StateTileState64 Word8 Word8 Word8 Word8 Word8 [HSBK] -- ^ 711
  -- Returned from a GetTileState64 and provides all the pixels in the
  -- specified rectangle for that tile.
  --
  -- Field       | Type
  -- tile_index  | unsigned 8-bit integer
  -- reserved    | unsigned 8-bit integer
  -- x           | unsigned 8-bit integer
  -- y           | unsigned 8-bit integer
  -- width       | unsigned 8-bit integer
  -- colors[64]  | 64 HSBK values

  | SetTileState64 Word8 Word8 Word8 Word8 Word8 Word8 Word32 [HSBK] -- ^ 715
  -- This lets you set 64 pixels from a starting x and y for a rectangle with the specified width.
  --
  -- For the LIFX Tile it really only makes sense to set x and y to zero, and width to 8.
  --
  -- Field       | Type
  -- tile_index  | unsigned 8-bit integer
  -- length      | unsigned 8-bit integer
  -- reserved    | unsigned 8-bit integer
  -- x           | unsigned 8-bit integer
  -- y           | unsigned 8-bit integer
  -- width       | unsigned 8-bit integer
  -- duration    | unsigned 32-bit integer
  -- colors[64]  | 64 HSBK values



src/Data/LIFX/Light.hs
-------------------------------------------------------------------------------
-- |
-- Module    :  Data.LIFX.Light
-- Copyright :  (c) Sam Stites 2017
-- License   :  BSD-3-Clause
-- Maintainer:  fnz@fgvgrf.vb (cipher:rot13)
-- Stability :  experimental
-- Portability: non-portable
-- Light Messages
--
-- Light messages control and acquire the state of a specific type of device,
-- such as a lightbulb. The state is comprised of the label, power level and
-- color.
--
-- There are a duplicate set of device power related messages,
-- Device::GetPower, Device::StatePower and Device::SetPower, which are
-- extended for lights by adding a duration field to Light::SetPower.
--
-- The Protocol header message type field value (decimal) appears after each
-- message name in this document. Each section below describes the payload that
-- is appended to the message header. Some messages do not require a payload.
-------------------------------------------------------------------------------
module Data.LIFX.Light where

import Prelude
import GHC.Word
import GHC.Int
import Data.ByteString

import Data.LIFX.Device

-- *  Message Field Data Types

-- | HSBK is used to represent the color and color temperature of a light.
--
-- The color is represented as an HSB (Hue, Saturation, Brightness) value.
--
-- The color temperature is represented in K (Kelvin) and is used to adjust the
-- warmness / coolness of a white light, which is most obvious when saturation
-- is close zero.
--
-- For user interface purposes, the hue is typically scaled to between 0° and
-- 360°. Saturation and brightness are typically scaled to between 0% and 100%.
data HSBK = HSBK
  { hue :: Word16             -- Hue: range 0 to 65535
  , saturation :: Word16      -- Saturation: range 0 to 65535
  , brightness :: Word16      -- Brightness: range 0 to 65535
  , kelvin :: Word16          -- Kelvin: range 2500° (warm) to 9000° (cool)
  } deriving (Eq, Ord, Show)


-- * Light Messages


data LightMessages
  = Get -- ^ 101
  -- Sent by a client to obtain the light state. No payload required. Causes the device to transmit a State message.

  | SetColor Word8 HSBK Word32 -- ^ 102
  -- Sent by a client to change the light state.
  --
  -- Field       | Type
  -- reserved    | unsigned 8-bit integer
  -- color       | HSBK
  -- duration    | unsigned 32-bit integer
  --
  -- The duration is the color transition time in milliseconds.
  --
  -- If the Frame Address res_required field is set to one (1) then the device will transmit a State message.

  | SetWaveform Word8 Word8 HSBK Word32 Float Int16 Word8 -- ^ 103
  -- Apply an effect to the bulb.
  --
  -- This message will be replied to with a State message.
  --
  -- Field       | Type                           | Purpose
  -- reserved    | unsigned 8-bit integer         |
  -- transient   | 8-bit integer as 0 or 1        | Color does not persist.
  -- color       | Hsbk                           | Light end color.
  -- period      | unsigned 32-bit integer        | Duration of a cycle in milliseconds.
  -- cycles      | 32-bit float                   | Number of cycles.
  -- skew_ratio  | signed 16-bit integer          | Waveform Skew, [-32768, 32767] scaled to [0, 1].
  -- waveform    | unsigned 8-bit integer         | Waveform to use for transition.
  --
  -- See Waveforms for more information.

  | SetWaveformOptional Word8 Bool HSBK Word32 Float Int16 Word8 Word8 Word8 Word8 Word8 -- ^ 119
  -- Optionally set effect parameters. Same as SetWaveform but allows some parameters to be set from the current value on device.
  --
  -- This message will be replied to with a State message.
  --
  -- Field           | Type
  -- reserved        | unsigned 8-bit integer
  -- transient       | boolean
  -- color           | Hsbk
  -- period          | unsigned 32-bit integer
  -- cycles          | 32-bit float
  -- skew_ratio      | signed 16-bit integer
  -- waveform        | unsigned 8-bit integer
  -- set_hue         | 8-bit integer as 0 or 1
  -- set_saturation  | 8-bit integer as 0 or 1
  -- set_brightness  | 8-bit integer as 0 or 1
  -- set_kelvin      | 8-bit integer as 0 or 1
  --
  -- See Waveforms for more information.

  | State HSBK Int16 PowerLevel Label Word64 -- ^ 107
  -- Sent by a device to provide the current light state.
  --
  -- Field       | Type
  -- color       | HSBK
  -- reserved    | signed 16-bit integer
  -- power       | unsigned 16-bit integer
  -- label       | string, size: 32 bytes
  -- reserved    | unsigned 64-bit integer
  --
  -- See color, label and power level.

  | GetPower -- ^ 116
  -- Sent by a client to obtain the power level. No payload required. Causes the device to transmit a StatePower message.

  | SetPower PowerLevel Word32 -- ^ 117
  -- Sent by a client to change the light power level.
  --
  -- Field       | Type
  -- level       | unsigned 16-bit integer
  -- duration    | unsigned 32-bit integer
  --
  -- The power level must be either 0 or 65535.
  --
  -- The duration is the power level transition time in milliseconds.
  --
  -- If the Frame Address res_required field is set to one (1) then the device will transmit a StatePower message.

  | StatePower PowerLevel -- ^ 118
  -- Sent by a device to provide the current power level.
  --
  -- Field       | Type
  -- level       | unsigned 16-bit integer

  | GetInfrared -- ^ 120
  -- Gets the current maximum power level of the Infrared channel. If the Frame Address res_required field is set to one (1) then the device will transmit a StateInfrared message.
  --
  -- No payload is required

  | StateInfrared Word16 -- ^ 121
  -- This message is returned from a GetInfrared message. It indicates the current maximum setting for the infrared channel.
  --
  -- Field       | Type
  -- brightness  | unsigned 16-bit integer

  | SetInfrared Word16 -- ^ 122
  -- Send this message to alter the current maximum brightness for the infrared channel.
  --
  -- If the Frame Address res_required field is set to one (1) then the device will transmit a StateInfrared message.
  --
  -- Field       | Type
  -- brightness  | unsigned 16-bit integer


src/Data/LIFX/Header.hs
-------------------------------------------------------------------------------
-- |
-- Module    :  Data.LIFX.Header
-- Copyright :  (c) Sam Stites 2017
-- License   :  BSD-3-Clause
-- Maintainer:  fnz@fgvgrf.vb (cipher:rot13)
-- Stability :  experimental
-- Portability: non-portable
--
-- All LIFX Protocol messages start with the header.
--
-- * Header Description
--
-- Each LIFX Protocol message has the following format
--
--     Frame | Frame Address | Protocol Header | Payload
--
-- The header is composed of the Frame, Frame Address and Protocol header. The
-- Payload is covered separately in the documentation for the various message
-- types.
-------------------------------------------------------------------------------
{-# LANGUAGE ScopedTypeVariables #-}
module Data.LIFX.Header
  ( Word(..)
  , Frame, frame
  , FrameAddress, frameAddress
  , ProtocolHeader, protocolHeader
  ) where

import Prelude hiding (Word, sequence)

import Codec.Serialise
import Codec.Serialise.Decoding
import Codec.Serialise.Encoding
import Data.Bits
import Data.ByteString (ByteString)
import Data.Coerce
import Control.Monad hiding (sequence)
import Data.Proxy
import GHC.Generics
import GHC.Word hiding (Word(..))
import GHC.TypeLits
import qualified Data.ByteString as BS

newtype Word (rep :: Nat) = Word Word64
  deriving (Eq, Ord, Show, Generic)

instance KnownNat r => Serialise (Word r) where
  encode :: forall r . KnownNat r => Word r -> Encoding
  encode (Word s) = foldMap (encodeBool . testBit s) [0..fromIntegral (natVal (Proxy @r)) - 1]

  decode :: forall r s . KnownNat r => Decoder s (Word r)
  decode = do
    bools <- replicateM (fromIntegral (natVal (Proxy @r))) decodeBool
    pure $ Word $ foldl (\r (i, b) -> if b then setBit r i else r) zeroBits $ zip [0..] bools

-- The Frame section contains information about the following:
--   Size of the entire message
--   LIFX Protocol number: must be 1024 (decimal)
--   Use of the Frame Address target field
--   Source identifier
--
-- Total bits: 16+2+1+1+12+32
data Frame = Frame
  { size        :: Word16          -- ^ Size of entire message in bytes including this field
  , origin      :: Word 2          -- ^ Message origin indicator: must be zero (0)
  , tagged      :: Bool            -- ^ Determines usage of the Frame Address target field
  , addressable :: Bool            -- ^ Message includes a target address: must be one (1)
  , protocol    :: Word 12         -- ^ Protocol number: must be 1024 (decimal)
  , source      :: Word32          -- ^ Source identifier: unique value set by the client, used by responses
  } deriving (Eq, Ord, Show, Generic)

frame
  :: (size ~ Word16, tagged ~ Bool, source ~ Word32)
  => size -> tagged -> source -> Frame
frame sz t = Frame sz (Word 0) t True (Word 1024)

instance Serialise Frame where
  encode :: Frame -> Encoding
  encode f
    =  encodeWord16 (size f)
    <> encode       (origin f)
    <> encodeBool   (tagged f)
    <> encodeBool   (addressable f)
    <> encode       (protocol f)
    <> encodeWord32 (source f)

  decode :: Decoder s Frame
  decode = Frame
    <$> decodeWord16
    <*> decode
    <*> decodeBool
    <*> decodeBool
    <*> decode
    <*> decodeWord32

-- The Frame Address section contains the following routing information:
--   Target device address
--   Acknowledgement message is required flag
--   State response message is required flag
--   Message sequence number
--
-- Total bits: 64+48+6+1+1+8
data FrameAddress = FrameAddress
  { target       :: Word64    -- ^ 6 byte device address (MAC address) or zero (0) means all devices
  , reserved0    :: Word 48   -- ^ Must all be zero (0)
  , reserved1    :: Word 6    -- ^ Reserved
  , ack_required :: Bool      -- ^ Acknowledgement message required
  , res_required :: Bool      -- ^ Response message required
  , sequence     :: Word8     -- ^ Wrap around message sequence number
  } deriving (Eq, Ord, Show)

frameAddress
  :: (target ~ Word64, ack_required ~ Bool, res_required ~ Bool, sequence ~ Word8)
  => target -> ack_required -> res_required -> sequence -> FrameAddress
frameAddress t a b c = FrameAddress t (Word 0) (Word 0) a b c

instance Serialise FrameAddress where
  encode :: FrameAddress -> Encoding
  encode f
    =  encodeWord64 (target       f)
    <> encode       (reserved0    f)
    <> encode       (reserved1    f)
    <> encodeBool   (ack_required f)
    <> encodeBool   (res_required f)
    <> encodeWord8  (sequence     f)

  decode :: Decoder s FrameAddress
  decode = FrameAddress
    <$> decodeWord64
    <*> decode
    <*> decode
    <*> decodeBool
    <*> decodeBool
    <*> decodeWord8

-- The Protocol header contains the following information about the message:
--   Message type which determines what action to take (based on the Payload)
data ProtocolHeader = ProtocolHeader
  { reserved2 :: Word64   -- ^ Reserved
  , mtype     :: Word16   -- ^ Message type determines the payload being used
  , reserved3 :: Word16   -- ^ Reserved
  } deriving (Eq, Ord, Show)

protocolHeader :: (Word16 ~ mtype) => mtype -> ProtocolHeader
protocolHeader t = ProtocolHeader 0 t 0

instance Serialise ProtocolHeader where
  encode :: ProtocolHeader -> Encoding
  encode f
    =  encodeWord64 (reserved2 f)
    <> encodeWord16 (mtype     f)
    <> encodeWord16 (reserved3 f)

  decode :: Decoder s ProtocolHeader
  decode = ProtocolHeader
    <$> decodeWord64
    <*> decodeWord16
    <*> decodeWord16

newtype Payload = Payload ByteString
  deriving (Eq, Ord, Show, Generic)




src/Data/LIFX/MultiZone.hs
-------------------------------------------------------------------------------
-- |
-- Module    :  Data.LIFX.MultiZone
-- Copyright :  (c) Sam Stites 2017
-- License   :  BSD-3-Clause
-- Maintainer:  fnz@fgvgrf.vb (cipher:rot13)
-- Stability :  experimental
-- Portability: non-portable
--
-- MultiZone Messages
-------------------------------------------------------------------------------
module Data.LIFX.MultiZone where

import Prelude
import GHC.Word
import GHC.Int
import Data.ByteString

import Data.LIFX.Device
import Data.LIFX.Light

-- * Enumerated Types

-- | This type allows you to provide hints to the device about how the changes
-- you make should be performed. For example you can send multiple zones and
-- have them all apply at once. Application Request is stored as an unsigned
-- 8-bit integer.
data ApplicationRequest
  = NO_APPLY   -- ^ Value: 0.  Don't apply the requested changes until a message with APPLY or APPLY_ONLY is sent.
  | APPLY      -- ^ Value: 1.  Apply the changes immediately and apply any pending changes.
  | APPLY_ONLY -- ^ Value: 2.  Ignore the requested changes in this message and only apply pending changes.

-- * Messages

data Messages
  = SetColorZones Word8 Word8 HSBK Word32 ApplicationRequest -- ^ 501
  -- This message is used for changing the color of either a single or
  -- multiple zones. The changes are stored in a buffer and are only applied
  -- once a message with either APPLY or APPLY_ONLY set.
  --
  -- Field       | Type
  -- start_index | unsigned 8-bit integer
  -- end_index   | unsigned 8-bit integer
  -- color       | HSBK
  -- duration    | unsigned 32-bit integer
  -- apply       | Application Request

  | GetColorZones Word8 Word8 -- ^ 502
  -- GetColorZones is used to request the zone colors for a range of zones.
  -- The bulb will respond with either StateZone or StateMultiZone messages as
  -- required to cover the requested range. The bulb may send state messages
  -- that cover more than the requested zones. Any zones outside the requested
  -- indexes will still contain valid values at the time the message was sent.
  --
  -- Field       | Type
  -- start_index | unsigned 8-bit integer
  -- end_index   | unsigned 8-bit integer

  | StateZone Word8 Word8 HSBK -- ^ 503
  -- The StateZone message represents the state of a single zone with the
  -- index field indicating which zone is represented. The count field
  -- contains the count of the total number of zones available on the device.
  --
  -- Field       | Type
  -- count       | unsigned 8-bit integer
  -- index       | unsigned 8-bit integer
  -- color       | HSBK

  | StateMultiZone Word8 Word8 HSBK HSBK HSBK HSBK HSBK HSBK HSBK HSBK  -- ^ 506
  -- The StateMultiZone message represents the state of eight consecutive
  -- zones in a single message. As in the StateZone message the count field
  -- represents the count of the total number of zones available on the
  -- device. In this message the index field represents the index of color[0]
  -- and the rest of the colors are the consecutive zones thus the index of
  -- the color[n] zone will be index + n.
  --
  -- Field       | Type
  -- count       | unsigned 8-bit integer
  -- index       | unsigned 8-bit integer
  -- color[0]    | HSBK
  -- color[1]    | HSBK
  -- color[2]    | HSBK
  -- color[3]    | HSBK
  -- color[4]    | HSBK
  -- color[5]    | HSBK
  -- color[6]    | HSBK
  -- color[7]    | HSBK


src/Data/LIFX/Device.hs
-------------------------------------------------------------------------------
-- |
-- Module    :  Data.LIFX.Device
-- Copyright :  (c) Sam Stites 2017
-- License   :  BSD-3-Clause
-- Maintainer:  fnz@fgvgrf.vb (cipher:rot13)
-- Stability :  experimental
-- Portability: non-portable
--
-- Device messages control and acquire the state of a device, independent of its
-- specific type. The state is composed of the label, port, power level, time,
-- firmware version and microcontroller statistics.
--
-- These messages are common to all LIFX devices, which may also implement
-- device specific messages, such as Light Messages.
--
-- The Protocol header message type field value (decimal) appears after each
-- message name in this document. Each section below describes the payload that
-- is appended to the message header. Some messages do not require a payload.
-------------------------------------------------------------------------------
module Data.LIFX.Device where

import Prelude

import Codec.Serialise
import Codec.Serialise.Decoding
import Codec.Serialise.Encoding
import Data.ByteString
import GHC.Int
import GHC.Generics
import GHC.Word

-- *  Enumerated Types

-- | Describes the services exposed by the device.
--
-- When a device is discovered the Service types and IP port are provided.
--
-- All values other than those documented above are reserved for future protocol expansion.
--
-- The LIFX Protocol utilizes UDP/IP for all messages covered by this documentation.
data Service
  = UDP             -- ^ UDP value is 1, it should be represented as an unsigned 8-bit integer
  deriving (Eq, Ord, Show)

instance Serialise Service where
  encode :: Service -> Encoding
  encode UDP = encodeBool True

  decode :: Decoder s Service
  decode =
    decodeBool >>= \case
      True -> pure UDP
      False -> fail "Error: Expected to decode a single bit-encoded 1 representing UDP. Found 0."

serviceValue :: Num i => Service -> i
serviceValue UDP = 1

-- *  Message Field Data Types

-- | For user interfaces purposes, each device can be identified by their label as well as a group and a location.
--
-- The label string is a fixed-length field, which is not null-terminated.
--
-- If you are setting a label/group/location for your light, it is highly recommended you store the label as a utf-8 string for compatibility with existing clients.
newtype Label = Label String
  deriving (Eq, Ord, Show, Generic)


instance Serialise Label

-- The IP port number used by a LIFX device for a specific 'Service'.
--
-- Whilst LIFX devices usually listen to port 56700, it is strongly recommended that devices are discovered and the provided port number is used.
--
-- As of LIFX Protocol V2, LIFX devices sending unicast message responses will reply to whichever port number the client has bound. See message header frame source identifier regarding the conditions under which unicast messages are sent.
--
-- Prior to LIFX Protocol V2, i.e the Original LIFX A21 lightbulb running firmware version 1.x, LIFX devices send broadcast message responses to port 56700.
--
-- To get the best compatibility across both newer and older LIFX devices, clients should bind to UDP port 56700.
newtype Port = Port Word32
  deriving (Eq, Ord, Show, Generic)

instance Serialise Port

-- | The power level can be either standby (0) or enabled (65535).
-- Currently, only the values 0 and 65535 are supported
data PowerLevel
  = Standby
  | Enabled
  deriving (Eq, Ord, Show, Generic)

instance Serialise PowerLevel where
  encode = \case
    Standby -> encodeWord16 0
    Enabled -> encodeWord16 65535

  decode =
    decodeWord16 >>= \case
      0     -> pure Standby
      65535 -> pure Enabled
      x -> fail $ "Error: Powerlevel only supports the values 0 (standby) and 65535 (enabled). Found: " ++ show x

-- | All time values have a precision of nanoseconds. When an absolute time value is provided, then it is the number of nanoseconds since the epoch, i.e Thursday 1st January 1970 00:00:00.
newtype Time = Time Word64
  deriving (Eq, Ord, Show, Generic)

instance Serialise Time

-- *  Device Messages

data DeviceMessages
  = GetService -- ^ 2
  -- Sent by a client to acquire responses from all devices on the
  -- local network. No payload is required. Causes the devices to transmit
  -- a 'StateService' message.
  --
  -- When using this message the Frame tagged field must be set to one (1).

  | StateService Service Port  -- ^ 3
  -- Response to GetService message.
  --
  -- Provides the device Service and port. If the Service is temporarily
  -- unavailable, then the port value will be 0.
  --
  -- Field0: service - unsigned 8-bit integer, maps to Service
  -- Field0: port    - unsigned 32-bit integer

  | GetHostInfo  -- ^ 12
  -- Get Host MCU information. No payload is required. Causes the device to transmit a StateHostInfo message.

  | StateHostInfo Float Word32 Word32 Int16 -- ^ 13
  -- Response to GetHostInfo message.
  --
  -- Provides host MCU information.
  --     signal: radio receive signal strength in milliWatts
  --     tx: bytes transmitted since power on
  --     rx: bytes received since power on
  --
  -- Field    | Type
  -- signal   | 32-bit float
  -- tx       | unsigned 32-bit integer
  -- rx       | unsigned 32-bit integer
  -- reserved | signed 16-bit integer

  | GetHostFirmware -- ^ 14
  -- Gets Host MCU firmware information. No payload is required. Causes the device to transmit a StateHostFirmware message.

  | StateHostFirmware Word64 Word64 Word32 -- ^ 15
  -- Response to GetHostFirmware message.
  --
  -- Provides host firmware information.
  --
  --     build: firmware build time (absolute time in nanoseconds since epoch)
  --     version: firmware version
  --
  -- Field          | Type
  -- build          | unsigned 64-bit integer
  -- reserved       | unsigned 64-bit integer
  -- version        | unsigned 32-bit integer

  | GetWifiInfo -- ^ 16
  -- Get Wifi subsystem information. No payload is required. Causes the device to transmit a StateWifiInfo message.

  | StateWifiInfo Float Word32 Word32 Int16  -- ^ 17
  -- Response to GetWifiInfo message.
  --
  -- Provides Wifi subsystem information.
  --
  --     signal: radio receive signal strength in mw
  --     tx: bytes transmitted since power on
  --     rx: bytes received since power on
  --
  -- Field       | Type
  -- signal      | 32-bit float
  -- tx          | unsigned 32-bit integer
  -- rx          | unsigned 32-bit integer
  -- reserved    | signed 16-bit integer

  | GetWifiFirmware Word64 Word64 Word32 -- ^ 18
  -- Get Wifi subsystem firmware. No payload is required. Causes the device to transmit a StateWifiFirmware message.
  -- StateWifiFirmware - 19
  --
  -- Response to GetWifiFirmware message.
  --
  -- Provides Wifi subsystem information.
  --
  --     build: firmware build time (absolute time in nanoseconds since epoch)
  --     version: firmware version
  --
  -- Field     | Type
  -- build     | unsigned 64-bit integer
  -- reserved  | unsigned 64-bit integer
  -- version   | unsigned 32-bit integer

  | GetPower -- ^ 20
  -- Get device power level. No payload is required. Causes the device to transmit a StatePower message.

  | SetPower PowerLevel -- ^ 21
  -- Set device power level.
  --
  -- Zero implies standby and non-zero sets a corresponding power draw level. Currently only 0 and 65535 are supported.
  -- Field          | Type
  -- level          | unsigned 16-bit integer

  | StatePower PowerLevel -- ^ 22
  -- Response to GetPower message.
  --
  -- Provides device power level.
  --
  -- Field          | Type
  -- level          | unsigned 16-bit integer

  | GetLabel -- ^ 23
  -- Get device label. No payload is required. Causes the device to transmit a StateLabel message.

  | SetLabel ByteString -- ^ 24
  -- Set the device label text.
  --
  -- Field          | Type
  -- label          | string, size: 32 bytes

  | StateLabel ByteString -- ^ 25
  -- Response to GetLabel message.
  --
  -- Provides device label.
  --
  -- Field          | Type
  -- label          | string, size: 32 bytes

  | GetVersion -- ^ 32
  -- Get the hardware version. No payload is required. Causes the device to transmit a StateVersion message.

  | StateVersion Word32 Word32 Word32 -- ^ 33
  -- Response to GetVersion message.
  --
  -- Provides the hardware version of the device. See LIFX Products for how to interpret the vendor and product ID fields.
  --
  --     vendor: vendor ID
  --     product: product ID
  --     version: hardware version
  --
  -- Field          | Type
  -- vendor         | unsigned 32-bit integer
  -- product        | unsigned 32-bit integer
  -- version        | unsigned 32-bit integer

  | GetInfo -- ^ 34
  -- Get run-time information. No payload is required. Causes the device to transmit a StateInfo message.

  | StateInfo Time Time Time -- ^ 35
  -- Response to GetInfo message.
  --
  -- Provides run-time information of device.
  --
  --     time: current time (absolute time in nanoseconds since epoch)
  --     uptime: time since last power on (relative time in nanoseconds)
  --     downtime: last power off period, 5 second accuracy (in nanoseconds)
  --
  -- Field       | Type
  -- time        | unsigned 64-bit integer
  -- uptime      | unsigned 64-bit integer
  -- downtime    | unsigned 64-bit integer

  | Acknowledgement -- ^ 45
  -- Response to any message sent with _ack_required_ set to 1. See message header frame address.

  | GetLocation -- ^ 48
  -- Ask the bulb to return its location information.
  -- No payload is required.
  -- Causes the device to transmit a StateLocation message.

  | SetLocation ByteString Label Time -- ^ 49
  -- Set the device location.
  --
  -- Applications wishing to change either the label or location attributes MUST
  -- set the updated_at field to the current timestamp and send the message to
  -- all applicable devices that are currently online. This is because when
  -- reading these values the applications will consider unique location
  -- fields to be a location identifier, and the label on the bulb with the
  -- highest updated_at field for that location will be used. Applications
  -- SHOULD attempt to correct any labels that are out of date when found.
  --
  --It is recommended to set the response_required flag on the message header
  --to speed up updating the cloud account.
  --
  -- When creating a new location generate the location field with random data.
  --
  --     location: guid byte array.
  --     label: text label for location.
  --     updated_at: UTC timestamp of last label update in nanoseconds.
  --
  -- Field          | Type
  -- location       | byte array, size: 16
  -- label          | string, size: 32
  -- updated_at     | integer, size: 64

  | StateLocation ByteString Label Time -- ^ 50
  -- Device location.
  --
  -- Field          | Type
  -- location       | byte array, size: 16 bytes
  -- label          | string, size: 32 bytes
  -- updated_at     | unsigned 64-bit integer

  | GetGroup -- ^ 51
  -- Ask the bulb to return its group membership information.
  -- No payload is required.
  -- Causes the device to transmit a StateGroup message.

  | SetGroup ByteString Label Time -- ^ 52
  -- Set the device group.
  --
  -- Applications wishing to change either the label or group attributes MUST
  -- set the updated_at field to the current timestamp and send the message to
  -- all applicable devices that are currently online. This is because when
  -- reading these values the applications will consider unique group fields
  -- to be a group identifier, and the label on the bulb with the highest
  -- updated_at field for that group will be used. Applications SHOULD attempt
  -- to correct any labels that are out of date when found.
  --
  -- It is recommended to set the response_required flag on the message header
  -- to speed up updating the cloud account.
  --
  -- When creating a new group generate the group field with random data.
  --
  --     group: guid byte array.
  --     label: text label for group.
  --     updated_at: UTC timestamp of last label update in nanoseconds.
  --
  -- Field          | Type
  -- group          | byte array, size: 16
  -- label          | string, size: 32
  -- updated_at     | integer, size: 64

  | StateGroup ByteString Label Time -- ^ 53
  -- Device group.
  --
  -- Field          | Type
  -- group          | byte array, size: 16 bytes
  -- label          | string, size: 32 bytes
  -- updated_at     | unsigned 64-bit integer

  | EchoRequest ByteString -- ^ 58
  -- Request an arbitrary payload be echoed back. Causes the device to transmit an EchoResponse message.
  --
  -- Field          | Type
  -- payload        | byte array, size: 64 bytes

  | EchoResponse ByteString -- ^ 59
  -- Response to EchoRequest message.
  --
  -- Echo response with payload sent in the EchoRequest.
  -- Field          | Type
  -- payload        | byte array, size: 64 bytes
